pkgs/swiftgen/swift2objc/Sources/swift2objc/AST/TypeSyntax.swift

```diff
@@ -0,0 +1,12 @@
+//
+//  File.swift
+//  
+//
+//  Created by Mohammad Eid on 13/05/2024.
+//
+
+import Foundation
+
+struct TypeSyntax {
```

Comment:
From my reading, it seems like you've got this `TypeSyntax` struct that represents any type by its name, and the `StructSyntax` struct is totally unrelated. So I guess the idea is that if you have a `TypeSyntax` whose name refers to a struct, you'd look up the corresponding `StructSyntax` in some sort of global map? If that's the approach you want to go with, I think the `name` alone will not be enough to uniquely identify all types (think about two types in different libs with the same name, or generic types that differ by their type args, or function types that have returns and args). You might need some concept of id instead.

The approach we use in ffigen is that we have a top level [`Type` class](https://github.com/dart-lang/native/blob/main/pkgs/ffigen/lib/src/code_generator/type.dart#L13), and a whole inheritance hierarchy of other types. For example, if you had a function that took a struct as a parameter, its type would be a [`FuncType` object](https://github.com/dart-lang/native/blob/main/pkgs/ffigen/lib/src/code_generator/func.dart#L40), with a [`Struct` object](https://github.com/dart-lang/native/blob/main/pkgs/ffigen/lib/src/code_generator/struct.dart#L31) as one of its args. That way there's one less layer of indirection to worry about, and it's a bit cleaner.

We do still need a global table of types in ffigen, to dedupe the types so that they're canonical (that is, we want it to be the case in ffigen that any two `Type`s that refer to the same type are identical by reference). If you're going to store your types in such a table, then they'll all need to inherit from some base class or implement the same protocol. And if you have to do that anyway, you may as well copy ffigen's approach. So make `TypeSyntax` a protocol that `StructSyntax` implements.

I also wonder whether all these AST objects should be `struct`s or `class`es. I guess anything that needs to be deduped, or might contain a cycle (eg a class with a method that returns an instance of that class), needs to be a `class` (so `StructSyntax`). Everything else can remain a struct.

Reply:
> `TypeSyntax` was supposed to be like a top level type that all other types extend.

Structs can't inherit from other structs in swift, they can only inherit from protocols. So `TypeSyntax` would need to be a protocol in that case.

> I didn't consider the need of having a global table though. I'll try to check out ffigen's approach tomorrow and see what I can do.

Ffigen is a bit hard to read and needs some refactoring, so it's probably not the best thing to base your implementation on. But the global table I'm talking about is the [`BindingsIndex`](https://github.com/dart-lang/native/blob/9d0207980a02429c23d55bf701508d01c94a0019/pkgs/ffigen/lib/src/header_parser/utils.dart#L386).

It's probably easiest to explain by example. You're walking over the parsed AST, looking for things you need to code-gen. You find a method that accepts a closure as an argument, so you want to generate a wrapper class for that closure type (assuming we're still doing that). But it turns out this same closure type is used in 5 different places, so now you have 5 copies of that wrapper class in your generated code. To avoid this, you have to dedupe your closure types.

There are a few ways of doing this, but the way ffigen does it is to store all the seen ObjC block types in the `BindingsIndex`. Every time it finds a block, it checks the index to see if it's been seen before. If it's new it creates an `ObjCBlock` AST node and stores it in the index, otherwise it returns the `ObjCBlock` already in the index. These AST nodes are all classes (dart doesn't have structs), so that means that all the `ObjCBlock`s of a particular type are the same object (modifying one modifies them all).

> Also I had a quick look at `SwiftSyntax` types declaration and apparently they're using structs too.

It looks like they're using the structs+protocols approach.

> > I guess anything that needs to be deduped, or might contain a cycle
> 
> I didn't quite understand what you meant there. Can you elaborate?

I talked more about deduping above. As for cycles, I just mean that since structs are pass by value, you can't have a `Foo` struct that contains a `Foo` struct. Pass by value means you get the whole struct as a direct block of memory, whereas with pass by reference means you get a reference that points to the block of memory. So when structs are included in other objects they are inlined into that object's memory (this is how it works in other languages, and I assume the same is true in swift). So if you had a `Foo` that contains a `Foo`, that `Foo` also needs to contain a `Foo`, which also needs an inline `Foo`, and so on. Your struct would get infinitely big.

So later when you add a `ClassSyntax` struct, and it needs to have a field for its super type, how is that going to be represented? The `ClassSyntax` struct can't contain a `ClassSyntax`.

It gets a bit more complicated in swift because when a struct inherits from a protocol they box the struct. So instead of storing the struct inline you're now storing a reference to a struct that lives elsewhere in memory. Presumably that's how `SwiftSyntax` is getting around the issue.

This super type example is also a case where deduping comes in. If two classes inherit from the same super class, their copy of the super class object should probably be identical. It's convenient, and also saves memory (you don't want each child class to store a separate copy of the entire chain of super classes). I think even the boxed structs you get if you use protocols are still copied around like that, so you can't really do deduping using structs+protocols.

> So I think we can use structs and use classes when needed.

Yeah, I think it makes sense to use structs for most of these smaller objects. At the moment I think the only thing that needs to be a class (or use the struct+protocol approach) is `TypeSyntax` and `StructSyntax`.

Reply:
> Structs can't inherit from other structs in swift, they can only inherit from protocols. So TypeSyntax would need to be a protocol in that case.

Yeah idk why I made it a struct to begin with. It was probably a "brainstorming" phase.

So I checked out `BindingsIndex` and I can see how this would work well with using classes for to represent `FooSyntax` entities. I think it's safe to use structs for things that hold primitive data and no complex objects. I'll be working on that next and let u know when it's done.